File: Sources/MoonPlanner/Plan.swift
import MoonAST

public enum DagNodeKind: String, Codable, Equatable, Sendable {
    case action
    case bind
    case mapM
    case mapM_join
    case storm
}

public struct DagNode: Codable, Equatable, Sendable {
    public var id: String
    public var kind: DagNodeKind
    public var label: String
    public var dependencies: [String]
    public var bindVar: String?
    public var mapMListVar: String?
    public var mapMFunc: String?
    public var stmtIndex: Int

    public init(
        id: String,
        kind: DagNodeKind,
        label: String,
        dependencies: [String],
        bindVar: String? = nil,
        mapMListVar: String? = nil,
        mapMFunc: String? = nil,
        stmtIndex: Int
    ) {
        self.id = id
        self.kind = kind
        self.label = label
        self.dependencies = dependencies
        self.bindVar = bindVar
        self.mapMListVar = mapMListVar
        self.mapMFunc = mapMFunc
        self.stmtIndex = stmtIndex
    }
}

public struct ExecutionDag: Codable, Equatable, Sendable {
    public var functionName: String
    public var nodes: [DagNode]

    public init(functionName: String, nodes: [DagNode]) {
        self.functionName = functionName
        self.nodes = nodes
    }
}

public func findFunction(_ program: Program, name: String) -> DoBlock? {
    for decl in program.declarations {
        guard case .function(let fnDecl, _) = decl else { continue }
        for eq in fnDecl.equations {
            if eq.name == name, case .doBlock(let block) = eq.body {
                return block
            }
        }
    }
    return nil
}

public func planFunction(_ program: Program, functionName: String = "main") -> ExecutionDag? {
    guard let block = findFunction(program, name: functionName) else { return nil }

    var nodes: [DagNode] = []
    var varDefs: [String: String] = [:]
    var counter = 0

    func addNode(
        kind: DagNodeKind,
        label: String,
        dependencies: [String],
        bindVar: String? = nil,
        mapMListVar: String? = nil,
        mapMFunc: String? = nil,
        stmtIndex: Int
    ) -> String {
        let id = "n\(counter)"
        counter += 1
        nodes.append(DagNode(
            id: id,
            kind: kind,
            label: label,
            dependencies: dependencies,
            bindVar: bindVar,
            mapMListVar: mapMListVar,
            mapMFunc: mapMFunc,
            stmtIndex: stmtIndex
        ))
        return id
    }

    for (i, stmt) in block.statements.enumerated() {
        let deps = collectDependencies(stmt, varDefs: varDefs)

        switch stmt {
        case .storm(let pattern, _, _, _):
            let bindName = patternBindName(pattern)
            let id = addNode(
                kind: .storm,
                label: bindName ?? "storm_\(i)",
                dependencies: deps,
                bindVar: bindName,
                stmtIndex: i
            )
            if let bindName { varDefs[bindName] = id }

        case .bind(let pattern, let expr, _, _):
            if let mapM = detectMapM(expr) {
                var mapDeps = deps
                if let listDep = varDefs[mapM.listVar], !mapDeps.contains(listDep) {
                    mapDeps.append(listDep)
                }

                let bindName = patternBindName(pattern)
                let mapId = addNode(
                    kind: .mapM,
                    label: bindName ?? "mapM_\(i)",
                    dependencies: mapDeps,
                    bindVar: bindName,
                    mapMListVar: mapM.listVar,
                    mapMFunc: exprLabel(mapM.funcExpr),
                    stmtIndex: i
                )

                let joinId = addNode(
                    kind: .mapM_join,
                    label: "\(mapId)_join",
                    dependencies: [mapId],
                    bindVar: bindName,
                    mapMListVar: mapM.listVar,
                    stmtIndex: i
                )

                if let bindName { varDefs[bindName] = joinId }
            } else {
                let bindName = patternBindName(pattern)
                let id = addNode(
                    kind: .bind,
                    label: bindName ?? "bind_\(i)",
                    dependencies: deps,
                    bindVar: bindName,
                    stmtIndex: i
                )
                if let bindName { varDefs[bindName] = id }
            }

        case .letBind(let pattern, let expr, _):
            let bindName = patternBindName(pattern)
            let id = addNode(
                kind: .action,
                label: exprLabel(expr),
                dependencies: deps,
                stmtIndex: i
            )
            if let bindName { varDefs[bindName] = id }

        case .action(let expr, _, _):
            _ = addNode(
                kind: .action,
                label: exprLabel(expr),
                dependencies: deps,
                stmtIndex: i
            )
        }
    }

    return ExecutionDag(functionName: functionName, nodes: nodes)
}

private func patternBindName(_ pattern: Pattern) -> String? {
    if case .pVar(let name, _) = pattern { return name }
    return nil
}

private func collectDependencies(_ stmt: DoStatement, varDefs: [String: String]) -> [String] {
    let expr: Expression
    var config: [ConfigItem] = []

    switch stmt {
    case .letBind(_, let e, _):
        expr = e
    case .storm(_, let input, let cfg, _):
        expr = input
        config = cfg
    case .bind(_, let e, let cfg, _):
        expr = e
        config = cfg
    case .action(let e, let cfg, _):
        expr = e
        config = cfg
    }

    var free = freeVarsExpr(expr)
    for item in config {
        free.formUnion(freeVarsExpr(item.value))
    }

    var deps: [String] = []
    for v in free.sorted() {
        if let dep = varDefs[v], !deps.contains(dep) {
            deps.append(dep)
        }
    }
    return deps
}

public func dagToGraph(_ dag: ExecutionDag) -> [String: [String]] {
    var graph: [String: [String]] = [:]
    for node in dag.nodes {
        graph[node.id] = node.dependencies
    }
    return graph
}

public func reachableFrom(_ dag: ExecutionDag, startIds: [String]) -> Set<String> {
    let graph = dagToGraph(dag)
    var visited = Set<String>()
    var stack = startIds

    while let id = stack.popLast() {
        if visited.contains(id) { continue }
        visited.insert(id)
        for dep in graph[id] ?? [] {
            stack.append(dep)
        }
    }

    return visited
}

public func nodesByLabel(_ dag: ExecutionDag, label: String) -> [DagNode] {
    dag.nodes.filter { $0.label == label || $0.bindVar == label }
}

File: legacy/packages/planner/src/plan.ts
import type { DoBlock, DoStatement, Expression, Program } from "@moon/ast";
import { freeVarsExpr } from "./free-vars";
import { detectMapM, exprLabel } from "./mapm";

export type DagNodeKind = "action" | "bind" | "mapM" | "mapM_join" | "storm";

export interface DagNode {
  id: string;
  kind: DagNodeKind;
  label: string;
  dependencies: string[];
  bindVar?: string;
  mapMListVar?: string;
  mapMFunc?: string;
  stmtIndex: number;
}

export interface ExecutionDag {
  functionName: string;
  nodes: DagNode[];
}

export function findFunction(program: Program, name: string): DoBlock | null {
  for (const decl of program.declarations) {
    if (decl.kind !== "Function") continue;
    for (const eq of decl.decl.equations) {
      if (eq.name === name && "statements" in eq.body) {
        return eq.body;
      }
    }
  }
  return null;
}

export function planFunction(program: Program, functionName = "main"): ExecutionDag | null {
  const block = findFunction(program, functionName);
  if (!block) return null;

  const nodes: DagNode[] = [];
  const varDefs = new Map<string, string>();
  let counter = 0;

  const addNode = (node: Omit<DagNode, "id">): string => {
    const id = `n${counter++}`;
    nodes.push({ id, ...node });
    return id;
  };

  for (let i = 0; i < block.statements.length; i++) {
    const stmt = block.statements[i]!;
    const deps = collectDependencies(stmt, varDefs);

    if (stmt.kind === "Storm") {
      const id = addNode({
        kind: "storm",
        label: stmt.pattern.kind === "PVar" ? stmt.pattern.name : `storm_${i}`,
        dependencies: deps,
        bindVar: stmt.pattern.kind === "PVar" ? stmt.pattern.name : undefined,
        stmtIndex: i,
      });
      if (stmt.pattern.kind === "PVar") {
        varDefs.set(stmt.pattern.name, id);
      }
      continue;
    }

    if (stmt.kind === "Bind") {
      const mapM = detectMapM(stmt.expr);
      if (mapM) {
        const listDep = varDefs.get(mapM.listVar);
        const mapDeps = listDep ? [...deps] : deps;
        if (listDep && !mapDeps.includes(listDep)) {
          mapDeps.push(listDep);
        }

        const mapId = addNode({
          kind: "mapM",
          label: stmt.pattern.kind === "PVar" ? stmt.pattern.name : `mapM_${i}`,
          dependencies: mapDeps,
          bindVar: stmt.pattern.kind === "PVar" ? stmt.pattern.name : undefined,
          mapMListVar: mapM.listVar,
          mapMFunc: exprLabel(mapM.func),
          stmtIndex: i,
        });

        const joinId = addNode({
          kind: "mapM_join",
          label: `${mapId}_join`,
          dependencies: [mapId],
          bindVar: stmt.pattern.kind === "PVar" ? stmt.pattern.name : undefined,
          mapMListVar: mapM.listVar,
          stmtIndex: i,
        });

        if (stmt.pattern.kind === "PVar") {
          varDefs.set(stmt.pattern.name, joinId);
        }
        continue;
      }

      const id = addNode({
        kind: "bind",
        label: stmt.pattern.kind === "PVar" ? stmt.pattern.name : `bind_${i}`,
        dependencies: deps,
        bindVar: stmt.pattern.kind === "PVar" ? stmt.pattern.name : undefined,
        stmtIndex: i,
      });

      if (stmt.pattern.kind === "PVar") {
        varDefs.set(stmt.pattern.name, id);
      }
      continue;
    }

    const id = addNode({
      kind: "action",
      label: exprLabel(stmt.expr),
      dependencies: deps,
      stmtIndex: i,
    });

    if (stmt.kind === "Let" && stmt.pattern.kind === "PVar") {
      varDefs.set(stmt.pattern.name, id);
    }
  }

  return { functionName, nodes };
}

function collectDependencies(
  stmt: DoStatement,
  varDefs: Map<string, string>,
): string[] {
  const expr = stmt.kind === "Let"
    ? stmt.expr
    : stmt.kind === "Storm"
      ? stmt.input
      : stmt.expr;
  const free = freeVarsExpr(expr);
  if (stmt.kind === "Storm" || stmt.kind === "Bind" || stmt.kind === "Action") {
    for (const item of stmt.config) {
      for (const v of freeVarsExpr(item.value)) free.add(v);
    }
  }
  const deps: string[] = [];

  for (const v of free) {
    const dep = varDefs.get(v);
    if (dep && !deps.includes(dep)) {
      deps.push(dep);
    }
  }

  return deps;
}

export function dagToGraph(dag: ExecutionDag): Record<string, string[]> {
  const graph: Record<string, string[]> = {};
  for (const node of dag.nodes) {
    graph[node.id] = [...node.dependencies];
  }
  return graph;
}

export function reachableFrom(dag: ExecutionDag, startIds: string[]): Set<string> {
  const graph = dagToGraph(dag);
  const visited = new Set<string>();
  const stack = [...startIds];

  while (stack.length > 0) {
    const id = stack.pop()!;
    if (visited.has(id)) continue;
    visited.add(id);
    for (const dep of graph[id] ?? []) {
      stack.push(dep);
    }
  }

  return visited;
}

export function nodesByLabel(dag: ExecutionDag, label: string): DagNode[] {
  return dag.nodes.filter((n) => n.label === label || n.bindVar === label);
}